File: LeetCode/Liang/59SpiralMatrixII.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        if (n == 0) {
            return vector<vector<int>>();
        }
        
        vector<vector<int>> grid(n, vector<int>(n, 0));
        int count = 0;
        int rowMin = 0, rowMax = n - 1, colMin = 0, colMax = n - 1;
        while (rowMin <= rowMax && colMin <= colMax) {
            for (int j = colMin; j <= colMax; ++j) {
                grid.at(rowMin).at(j) = ++count;
            }
            ++rowMin;
            
            for (int i = rowMin; i <= rowMax; ++i) {
                grid.at(i).at(colMax) = ++count;
            }
            --colMax;
            
            for (int j = colMax; j >= colMin; --j) {
                grid.at(rowMax).at(j) = ++count;
            }
            --rowMax;
            
            for (int i = rowMax; i >= rowMin; --i) {
                grid.at(i).at(colMin) = ++count;
            }
            ++colMin;
        }
        return grid;
    }
};

File: LeetCode/Kai/59SpiralMatrixII.java
public class Solution {
    public int[][] generateMatrix(int n) {
        int[][] result = new int[n][n];
        int rowMin = 0;
        int rowMax = n-1;
        int colMin = 0;
        int colMax = n-1;
        int num = 1;
        
        while(true) {
            for(int i=colMin; i<=colMax; i++) {
                result[rowMin][i] = num++;
            }
            if(num>n*n) return result;
            rowMin++;
            
            for(int i=rowMin; i<=rowMax; i++) {
                result[i][colMax] = num++;
            }
            if(num>n*n) return result;
            colMax--;
            
            for(int i=colMax; i>=colMin; i--) {
                result[rowMax][i] = num++;
            }
            if(num>n*n) return result;
            rowMax--;
            
            for(int i=rowMax; i>=rowMin; i--) {
                result[i][colMin] = num++;
            }
            if(num>n*n) return result;
            colMin++;
        }
    }
}